File: CuiYanling/Maximal Rectangle.java
public class Solution {
    public int maximalRectangle(char[][] matrix) {
        int m = matrix.length;
        if(m==0) return 0;
        int n = matrix[0].length;
        int[][] height = new int[m][n];
        for(int i=0;i<m;i++)
            for(int j=0;j<n;j++)
                if(matrix[i][j]=='1'){
                    height[i][j]=1;
                    if(i>0&&height[i-1][j]>0)
                        height[i][j] = height[i-1][j]+1;
                }
        int ans = 0;
        for(int i=0;i<m;i++){
            int stack[] = new int[n+1];
            int top = -1;
            for(int j=0;j<=n;j++){
                int now = -1;
                if(j<n) now = height[i][j];
                if(top>=0&&now<=height[i][stack[top]]){
                    int index = stack[top];
                    while(top>=0&&now<=height[i][stack[top]]){
                        int width = index+1;
                        if(top>0) width = index-stack[top-1];
                        ans = Math.max(ans,height[i][stack[top]]*width);
                        --top;
                    }
                }
                stack[++top] = j;
            }
        }
        return ans;
    }
}

File: LuoSonglei/week-2/MaximalRectangle.c
/*******************************************
Author      : LHearen
E-mail      : LHearen@126.com
Time        : 2015-12-19 11:19
Description : 
Source      : https://leetcode.com/problems/maximal-rectangle/
*******************************************/
//AC - 4ms
int fixedMaximalRectangle(int* heights, int* stack, int size)
{
    int current = 0, max = 0; 
    int i = 0, j = 0;
    int top = -1;
    stack[top] = -1;
    for(; i < size; i++)
    {
        while(top > 0 && heights[stack[top]] > heights[i])
        {
            current = (i - stack[top - 1] - 1) * heights[stack[top]];
            top--;
            if(current > max)
                max = current;
        }
        stack[++top] = i;
    }
    return current > max? current : max;
}
int maximalRectangle(char** matrix, int rowSize, int colSize)
{
    colSize++;
    int* heights = (int*)malloc(sizeof(int) * colSize);
    int* stack = (int*)malloc(sizeof(int) * colSize);
    int i = 0, j = 0;
    int max = 0, current = 0; 
    for(j = 0; j < colSize; j++) 
    {
        heights[j] = 0;
        stack[j] = 0;
    }
    for(i = 0; i < rowSize; i++)
    {
        for(j = 0; j < colSize - 1; j++)
        {
            matrix[i][j] == '0' ? heights[j] = 0 : heights[j]++;
        }
        current = fixedMaximalRectangle(heights, stack, colSize);
        if(current > max)
           max = current; 
    }
    return max;
}
